ARROW-1568: [C++] Implement Drop Null  Kernel for Arrays

aocsa opened the pull request:
Implement "drop null" kernels that return array without nulls

Note: This can be implemented as a arrow::compute::VectorFunction because the size of the array is changed, so this function is not valid in a SQL-like context

aocsa commented:
This PR is ready to review :)

bkietz commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,139 @@ class TakeMetaFunction : public MetaFunction {
+  arrow::NumericBuilder<arrow::Int32Type> builder(memory_pool);
+  int64_t relative_index = 0;
+  for (int64_t chunk_index = 0; chunk_index < chunks->num_chunks(); ++chunk_index) {
+    auto column_chunk = chunks->chunk(chunk_index);
```
Unfortunately, MSVC is opinionated about implicit integral conversions:

https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/40193996/job/q88mm6wio371o3nc#L1016

```suggestion
    auto column_chunk = chunks->chunk(static_cast<int>(chunk_index));
```

There are a few similar ~~warning~~errors here which will need to be addressed before CI will be entirely green

bkietz commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,144 @@ class TakeMetaFunction : public MetaFunction {
+    const std::shared_ptr<Array>& column, MemoryPool* memory_pool) {
+  std::shared_ptr<arrow::Array> indices;
+  arrow::NumericBuilder<arrow::Int32Type> builder(memory_pool);
+  builder.Reserve(column->length() - column->null_count());
```
Every function which returns Status or Result<> will prompt a similar compiler error if not explicitly handled, since it could be ignoring (for example) an OOM Status https://github.com/apache/arrow/pull/10802/checks?check_run_id=3202879387#step:5:608
```suggestion
  RETURN_NOT_OK(builder.Reserve(column->length() - column->null_count()));
```

nirandaperera commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,144 @@ class TakeMetaFunction : public MetaFunction {
+// ----------------------------------------------------------------------
+// DropNull Implementation
+
+Result<std::shared_ptr<arrow::Array>> GetNotNullIndices(
+    const std::shared_ptr<Array>& column, MemoryPool* memory_pool) {
```
I feel like it would be nice to have the following signature
```c++
Status GetNotNullIndices( const std::shared_ptr<Array>& column, MemoryPool* memory_pool, std::shared_ptr<arrow::Array>* out_array);
```

bkietz commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,144 @@ class TakeMetaFunction : public MetaFunction {
+Result<std::shared_ptr<arrow::Array>> GetNotNullIndices(
+    const std::shared_ptr<Array>& column, MemoryPool* memory_pool) {
+  std::shared_ptr<arrow::Array> indices;
+  arrow::NumericBuilder<arrow::Int32Type> builder(memory_pool);
```
I think instead of explicitly building indices of valid bits, it'd be more compact and performant to reuse GetTakeIndices. You can zero-copy construct a BooleanArray which is true where `column` is valid and false where `column` is null, then pass it to GetTakeIndices to acquire `indices`

westonpace replied:
If you had such a boolean array (and this may be future work) could it be more efficient to have a take that accepts a boolean array (of the same size as the input) and doesn't even construct the intermediate indices array?

nirandaperera commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,144 @@ class TakeMetaFunction : public MetaFunction {
+  return indices;
+}
+
+Result<std::shared_ptr<RecordBatch>> DropNullRecordBatch(const RecordBatch& batch,
```
I'm a little confused on the sematics for `DropNull` in record batches. What would happen in the following case?

| A | B    | C    |
|---|------|------|
| 1 | 4    | null |
| 2 | null | 6    |
| 3 | 5    | null |

nirandaperera replied:
I have a feeling that we need to make a union out of all the validity buffers and then call `Take` based on that?

aocsa replied:
DropNullRecordBatch now uses internally GetTakeIndicesImpl and it takes into consideration the offset.

nirandaperera replied:
well, it's not actually about the offset. What would be the output of the above example? An empty table?

aocsa replied:
Yes, the output would be an empty table. I am following the pandas dropna API
https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dropna.html

nirandaperera commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,144 @@ class TakeMetaFunction : public MetaFunction {
+    ("The output is populated with values from the input without the null values"),
+    {"input"});
+
+class DropNullMetaFunction : public MetaFunction {
```
There's an alternative approach to using `Take` (at least for arrays). We can promote array's validity buffer to a BooleanArray and call `Filter`. In that case, we dont have to create the valid indices IMO

nirandaperera commented on cpp/src/arrow/compute/kernels/vector_selection_test.cc:
```diff
@@ -1718,5 +1718,240 @@ TEST(TestTake, RandomFixedSizeBinary) {
+    CheckDropNull(type_singleton(), values, expected);
+  }
+
+  std::shared_ptr<DataType> type_singleton() {
```
nit: you can keep this as a class member

nirandaperera reviewed (commented):
I made some comments on the general approach. Great work @aocsa :clap:

westonpace commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,144 @@ class TakeMetaFunction : public MetaFunction {
+}
+
+Result<std::shared_ptr<Table>> DropNullTable(const Table& table, ExecContext* ctx) {
+  if (table.num_rows() == 0) {
```
If this check makes sense here then should we have the same check for the record batch and array versions?

aocsa replied:
Sure makes sense.

westonpace commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,144 @@ class TakeMetaFunction : public MetaFunction {
+      relative_index += column_chunk->length();
+    }
+  }
+  std::shared_ptr<arrow::Array> indices;
```
Can you add a check here.  If the length of `notnull_indices` is equal to the length of the table then you can just return the table an save some time.

westonpace commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,144 @@ class TakeMetaFunction : public MetaFunction {
+    switch (args[0].kind()) {
+      case Datum::ARRAY: {
+        const auto values = args[0].make_array();
+        ARROW_ASSIGN_OR_RAISE(auto indices,
```
If `values.null_count` is known (and 0) can we skip some work here?

westonpace commented on cpp/src/arrow/compute/kernels/vector_selection_test.cc:
```diff
@@ -1718,5 +1718,240 @@ TEST(TestTake, RandomFixedSizeBinary) {
+  }
+};
+
+TYPED_TEST_SUITE(TestDropNullKernelWithNumeric, NumericArrowTypes);
```
Since the implementation of DropNull relies on the Take kernel and there aren't different versions of it for each data type I don't know how much we need to test drop null on all the different types.

@nirandaperera / @bkietz I'd be interested in a second opinion here as I haven't spent a lot of time in kernels code.

westonpace commented on cpp/src/arrow/compute/kernels/vector_selection_test.cc:
```diff
@@ -1718,5 +1718,240 @@ TEST(TestTake, RandomFixedSizeBinary) {
+    ])";
+  CheckDropNull(union_type, union_json, union_json);
+}
+
```
I don't know what the rules are for what parts get tested in C++ and what parts get tested in JSON.  @nirandaperera / @bkietz maybe you can help me out here.  However, there are no C++ tests for the record batch, chunked array, or table versionsof drop null.  Though maybe this is ok.

nirandaperera replied:
AFAIK, we have utils build for Scalar kernels. But not many are available for vector kernels (and one main reason is, in vector kernels there are a lot of moving pieces). 

I'd like to suggest that we add a test case with random data (say 1000 rows), so that we know we go beyond the byte boundaries (for bitmaps), etc.

aocsa replied:
Test for record batch, chunked array and table are covered in the python test. IMO the c++ tests should be covered in another PR.

westonpace commented on python/pyarrow/tests/test_compute.py:
```diff
@@ -974,6 +974,51 @@ def test_take_null_type():
+
+
+def test_dropnull_table():
+    table = pa.table([pa.array(["a", None, "c", "d", None])], names=["a"])
```
Can you make some tests for tables with multiple columns?

westonpace commented on python/pyarrow/tests/test_compute.py:
```diff
@@ -974,6 +974,51 @@ def test_take_null_type():
+
+def test_dropnull_record_batch():
+    batch = pa.record_batch(
+        [pa.array(["a", None, "c", "d", None])], names=["a'"])
```
Can you make some tests for record batches with multiple columns? (I think this will fail if they have different numbers of null elements).

westonpace reviewed (changes requested):
I agree with Niranda that the record batch implementation probably won't work but I think this is a great start and looks like a very good addition.  Thanks for making this!

aocsa commented:
I updated this PR base on last feedback comments. Main change, the way algorithm drop nulls now are based on filter operation and  boolean filter array is obtained directly using the bitmap from input array.

nirandaperera commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,182 @@ class TakeMetaFunction : public MetaFunction {
+                         std::shared_ptr<arrow::BooleanArray>* out_array) {
+  auto bitmap_buffer = values.null_bitmap();
+  *out_array = std::make_shared<BooleanArray>(values.length(), bitmap_buffer, nullptr,
+                                              values.null_count(), values.offset());
```
```suggestion
                                              0, values.offset());
```
Once the validity buffer is promoted to a data buffer, we should consider it as a non-null array.   `values.null_count() != 0`, so, we can simply pass 0 here.

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,185 @@ class TakeMetaFunction : public MetaFunction {
+
+const FunctionDoc dropnull_doc(
+    "DropNull kernel",
+    ("The output is populated with values from the input without the null values"),
```
Given that this also supports RecordBatches/Tables, I think it would be good to explicitly describe the behaviour in those cases (as there are multiple options, eg the any/all in pandas).

jorisvandenbossche commented on python/pyarrow/compute.py:
```diff
@@ -590,6 +590,38 @@ def take(data, indices, *, boundscheck=True, memory_pool=None):
     return call_function('take', [data, indices], options, memory_pool)
 
 
+def dropnull(data, *, memory_pool=None):
```
Is this explicit function needed? (the auto-generated function doesn't work, or are the vector functions not auto-generated?)

aocsa replied:
To the best of my knowledge, this function is needed to be exposed as part of `pyarrow.compute` module https://github.com/apache/arrow/blob/082b11e9ecb4acd60d2e36ba8b5b3bda8b2027ce/python/pyarrow/lib.pyx#L113

jorisvandenbossche replied:
I quickly tried it out, and also without this explicit function, the auto-generated one seems to work fine. 

cc @pitrou do you know on what basis it's decided when to use the auto-generated one and when to add an explicit function?

pitrou replied:
The explicit functions are mostly historical (auto-generation became less and less limited over time), but in some cases they allow for more user-friendly signatures. If your explicit function has the same signature as the auto-generated one, then it's basically useless :-)

jorisvandenbossche replied:
I think the explicit function can thus be removed.

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,185 @@ class TakeMetaFunction : public MetaFunction {
+
+class DropNullMetaFunction : public MetaFunction {
+ public:
+  DropNullMetaFunction() : MetaFunction("dropnull", Arity::Unary(), &dropnull_doc) {}
```
```suggestion
  DropNullMetaFunction() : MetaFunction("drop_null", Arity::Unary(), &dropnull_doc) {}
```

(I think we generally use underscores, eg we already have `"is_null"` kernel)

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,185 @@ class TakeMetaFunction : public MetaFunction {
+}
+
+const FunctionDoc dropnull_doc(
+    "DropNull kernel",
```
```suggestion
    "Drop nulls from the input",
```

jorisvandenbossche reviewed (commented):
Cool, added a few comments!

How does "drop_null" work for nested types? I suppose it only takes into account the top-level validity? (if so, might be worth to test (if not yet done) and document explicitily)

aocsa commented:
I updated this PR addressing latest feedback comments.
I think this PR is ready for merging. Let me know if anything needs to be taken care of from my end. cc @nirandaperera, @jorisvandenbossche, @westonpace

nirandaperera commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2146,184 @@ class TakeMetaFunction : public MetaFunction {
+      ::arrow::internal::BitmapAnd(column->null_bitmap_data(), column->offset(),
+                                   dst->data(), 0, column->length(), 0,
+                                   dst->mutable_data());
+    } else {
+      for (int64_t i = 0; i < column->length(); ++i) {
+        if (!column->IsValid(i)) {
+          BitUtil::ClearBit(dst->mutable_data(), i);
+        }
```
I think this else block is redundant. `column->null_bitmap_data() == nullptr` means that all are valid. So we can omit this branch!

nirandaperera commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2146,184 @@ class TakeMetaFunction : public MetaFunction {
+}
+
+Result<std::shared_ptr<RecordBatch>> DropNullRecordBatch(const RecordBatch& batch,
+                                                         ExecContext* ctx) {
```
Shall we have an early termination here, that checks if none of the columns have nulls, simply return the input?

nirandaperera commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2146,184 @@ class TakeMetaFunction : public MetaFunction {
+  }
+  const int num_columns = table.num_columns();
+  std::vector<ArrayVector> inputs(num_columns);
+
```
we can have an early termination for  tables as well IMO (if all chunkedarrays are non-null, just return input)

nirandaperera commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2146,184 @@ class TakeMetaFunction : public MetaFunction {
+         ++chunk_index) {
+      const auto& column_chunk = inputs[col][chunk_index];
+      for (int64_t i = 0; i < column_chunk->length(); ++i) {
+        if (!column_chunk->IsValid(i)) {
+          BitUtil::ClearBit(dst->mutable_data(), relative_index + i);
```
I think you can use the `BitmapAnd` approach you used in record batches here. Only thing is you'll have to manipulate the offsets and lengths properly :slightly_smiling_face:

aocsa replied:
Thanks @nirandaperera. I updated the PR based on latest feedback. Nice to know more about the Arrow BitUtils :)

nirandaperera reviewed (changes requested):
Great work @aocsa :+1: I left some minor comments :-)

aocsa commented:
Thanks @nirandaperera. I updated the PR based on latest feedback. Nice to know more about the Arrow BitUtils :)
IMO this PR is ready to be merged :) +1 cc @westonpace

westonpace commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,203 @@ class TakeMetaFunction : public MetaFunction {
+  return Status::OK();
+}
+
+Status CreateEmptyArray(std::shared_ptr<DataType> type, MemoryPool* memory_pool,
+                        std::shared_ptr<Array>* output_array) {
+  std::unique_ptr<ArrayBuilder> builder;
+  RETURN_NOT_OK(MakeBuilder(memory_pool, type, &builder));
+  RETURN_NOT_OK(builder->Resize(0));
+  ARROW_ASSIGN_OR_RAISE(*output_array, builder->Finish());
+  return Status::OK();
+}
```
This seems like a very useful utility function.  I wonder if we want to put it somewhere more visible.  There could be a CreateEmptyRecordBatch and CreateEmptyTable too.  Can be done in a follow-up PR.

nirandaperera replied:
Yep! We just had this discussion in Zulip. 
https://ursalabs.zulipchat.com/#narrow/stream/271283-help.2Fc.2B.2B/topic/creating.20empty.20array.2F.20table.20using.20Slice

https://issues.apache.org/jira/browse/ARROW-7051

aocsa replied:
Yeap, I will create an issue on Jira

aocsa replied:
https://issues.apache.org/jira/browse/ARROW-13579

aocsa replied:
Nice!

nirandaperera commented on cpp/src/arrow/compute/kernels/vector_selection_test.cc:
```diff
@@ -1734,5 +1734,240 @@ TEST(TestTake, RandomFixedSizeBinary) {
   TakeRandomTest<FixedSizeBinaryType>::Test(fixed_size_binary(16));
 }
 
+// ----------------------------------------------------------------------
```
I think we need to add following test cases
1. 0-lengthed inputs  (to test the early termination code paths)
2. non-zero but all null values on inputs  (to test the early termination code paths)
3. larger test case with more random values (say 1000 or so) --> Because bitmap ops work on 8 elements together.  So, we need to test beyond 8 elements in an array. 
ex: https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/scalar_if_else_test.cc#L56
4. also, it would be nicer to have test with non-zero offsets in the inputs (only for array impl)

westonpace commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,203 @@ class TakeMetaFunction : public MetaFunction {
+  return result.record_batch();
+}
+
+using ::arrow::internal::Bitmap;
```
Nit: Put this in the method or at the top of the namespace declaration.

westonpace commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,203 @@ class TakeMetaFunction : public MetaFunction {
+    BitUtil::SetBitsTo(concatenated_bitmap->mutable_data(), 0, table.num_rows(), true);
+
+    for (auto bitmap : bitmaps) {
+      if (bitmap.buffer()->data()) {
+        ::arrow::internal::CopyBitmap(bitmap.buffer()->data(), bitmap.offset(),
+                                      bitmap.length(),
+                                      concatenated_bitmap->mutable_data(), bitmap_offset);
+      }
```
This can be done in a follow-up but I wonder if this copy could be avoided by passing offsets to BitmapAnd?

aocsa replied:
Yes It cant be done avoiding copy +1

aocsa replied:
CopyBitmap can't be skipped for now.  BitmapAnd only has an unaligned implemnentation with respect to the offset (I don't why). So if chunk->length() % 8 != 0 BitmapAnd will fail. I would suggest to create a follow-up issue for this.

westonpace commented on python/pyarrow/tests/test_compute.py:
```diff
@@ -974,6 +974,83 @@ def test_take_null_type():
+
+
+def test_drop_null_chunked_array():
+    arr = pa.chunked_array([["a", None], ["c", "d", None]])
+    expected_drop = pa.chunked_array([["a"], ["c", "d"]])
```
```suggestion
    arr = pa.chunked_array([["a", None], ["c", "d", None], [None], []])
    expected_drop = pa.chunked_array([["a"], ["c", "d"], [], []])
```

westonpace commented on python/pyarrow/tests/test_compute.py:
```diff
@@ -974,6 +974,83 @@ def test_take_null_type():
+    result = batch.drop_null()
+    expected = pa.record_batch(
+        [pa.array(["c"]), pa.array(["c"])], names=["a'", "b'"])
+    print(result["a'"])
+    print(expected["a'"])
```
Remove these prints

westonpace reviewed (changes requested):
Only a few very minor things.  Most are fine for a follow up, remove the prints and I'll be good.

aocsa commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,203 @@ class TakeMetaFunction : public MetaFunction {
+    return Table::Make(table.schema(), empty_table, 0);
+  }
+
+  ARROW_ASSIGN_OR_RAISE(auto dst,
+                        AllocateEmptyBitmap(table.num_rows(), ctx->memory_pool()));
+  BitUtil::SetBitsTo(dst->mutable_data(), 0, table.num_rows(), true);
+
+  for (int col_index = 0; col_index < num_columns; ++col_index) {
+    const ArrayVector& chunks = table.column(col_index)->chunks();
+    std::vector<Bitmap> bitmaps(chunks.size());
+    for (size_t chunk_index = 0; chunk_index < chunks.size(); ++chunk_index) {
+      const auto& column_chunk = chunks[chunk_index];
+      bitmaps[chunk_index] = Bitmap(column_chunk->null_bitmap_data(),
+                                    column_chunk->offset(), column_chunk->length());
+    }
+    int64_t bitmap_offset = 0;
+    ARROW_ASSIGN_OR_RAISE(auto concatenated_bitmap,
+                          AllocateEmptyBitmap(table.num_rows(), ctx->memory_pool()));
+    BitUtil::SetBitsTo(concatenated_bitmap->mutable_data(), 0, table.num_rows(), true);
+
+    for (auto bitmap : bitmaps) {
+      if (bitmap.buffer()->data()) {
+        ::arrow::internal::CopyBitmap(bitmap.buffer()->data(), bitmap.offset(),
+                                      bitmap.length(),
+                                      concatenated_bitmap->mutable_data(), bitmap_offset);
+      }
+      bitmap_offset += bitmap.length();
+    }
+    ::arrow::internal::BitmapAnd(concatenated_bitmap->data(), 0, dst->data(), 0,
+                                 table.num_rows(), 0, dst->mutable_data());
+  }
```
Nice! Thanks Niranda

aocsa replied:
This change can't be done. `BitmapAnd` only has an `unaligned`  implemnentation with respect to the offset (I don't why). So if `chunk->length() % 8 != 0` BitmapAnd will fail. One way to do this is to create an aligned and concatenated bitmap for each column and them apply the bytewise BitmapAnd.

bkietz commented on cpp/src/arrow/compute/kernels/vector_selection_test.cc:
```diff
@@ -1734,5 +1734,372 @@ TEST(TestTake, RandomFixedSizeBinary) {
+  }
+};
+
+TYPED_TEST_SUITE(TestDropNullKernelWithString, BinaryTypes);
```
https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/40309444/job/ccnm03ackeryl0m8#L1119

```suggestion
TYPED_TEST_SUITE(TestDropNullKernelWithString, BinaryArrowTypes);
```

These were renamed in #10862

aocsa commented:
I updated the PR addressing feedback. 

The following test cases were added.
- 0-lengthed inputs (to test the early termination code paths)
- non-zero but all null values on inputs (to test the early termination code paths)
- Larger test case with more random values (say 1000 or so) --> Because bitmap ops work on 8 elements together. So, we need to test beyond 8 elements in an array.
- Test with non-zero offsets in the inputs (for chunkedarrays and tables)

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2273,27 +2311,20 @@ Result<std::shared_ptr<Table>> DropNullTable(const Table& table, ExecContext* ct
-    "DropNull kernel",
-    ("The output is populated with values from the input without the null values"),
+const FunctionDoc drop_null_doc(
+    "Drop Null kernel",
```
```suggestion
    "Drop nulls from the input",
```

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,216 @@ class TakeMetaFunction : public MetaFunction {
+const FunctionDoc drop_null_doc(
+    "Drop Null kernel",
+    ("The output is populated with values from the input (Array, ChunkedArray, "
+     "RecordBatch, or Table) without the null values"),
```
I think it would be good to explicitly describe the behaviour for the RecordBatch/Table cases, i.e. that it drops the full row if there is any null (as there are multiple options, it could also have been "all" instead of "any").

westonpace commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,219 @@ class TakeMetaFunction : public MetaFunction {
+    ("The output is populated with values from the input (Array, ChunkedArray, "
+     "RecordBatch, or Table) without the null values."
+     "Note that for the RecordBatch/Table cases, `drop_null` drops the full row if there "
+     "is any null"
+     "null."),
```
```suggestion
     "is any null."),
```

westonpace reviewed (approved):
This looks good to me.  @bkietz @nirandaperera did you want to review further?

jorisvandenbossche reviewed (changes requested):
@aocsa with your latest push, it seems you (accidentally?) undid some previous changes, eg the name "drop_null" went back the "dropnull" (not sure if there are others). 
(without adding new commits with your latest changes (but amending them to previous ones), it's difficult to see what actually changed)

Also, you resolved my two small doc comments of yesterday, but it doesn't seem something actually changed on those lines?

aocsa commented:
> @aocsa with your latest push, it seems you (accidentally?) undid some previous changes, eg the name "drop_null" went back the "dropnull" (not sure if there are others).
> (without adding new commits with your latest changes (but amending them to previous ones), it's difficult to see what actually changed)
> 
> Also, you resolved my two small doc comments of yesterday, but it doesn't seem something actually changed on those lines?

Yeap It was an accident.  I change back the change. Thanks for notice this. cc @jorisvandenbossche

nirandaperera commented:
LGTM. +1 @aocsa Thank you for doing this! :-)

aocsa commented:
> LGTM. +1 @aocsa Thank you for doing this! :-)

cc @westonpace

pitrou commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,219 @@ class TakeMetaFunction : public MetaFunction {
+    }
+  }
+  if (null_count == 0) {
+    return Table::Make(table.schema(), table.columns(), table.num_rows());
```
Better to return the original table, here and above.

pitrou replied:
Same comment again as for the record batch variant.

pitrou commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,219 @@ class TakeMetaFunction : public MetaFunction {
+    int64_t global_offset = 0;
+    ARROW_ASSIGN_OR_RAISE(auto concatenated_bitmap,
+                          AllocateEmptyBitmap(table.num_rows(), ctx->memory_pool()));
+    BitUtil::SetBitsTo(concatenated_bitmap->mutable_data(), 0, table.num_rows(), true);
```
Another possibility entirely is to get `TableBatchReader` to get a vector of record batches spanning the table, then call drop_null on each record batch and then `Table::FromRecordBatches` to get the resulting table.

pitrou replied:
Ping on this?

aocsa replied:
I followed your last suggest using `TableBatchReader`. The reason that I needed to concatenate all source bitmap chunks was that `BitmapAnd` only has an unaligned implementation with respect to the offset (I don't why) not the bitmap length. So if `chunk->length() % 8 != 0` `BitmapAnd` will fail. https://github.com/apache/arrow/blob/820e5061847c9d6d261c416e57d6013321175565/cpp/src/arrow/util/bitmap_ops.cc#L285

Referenced before
https://github.com/apache/arrow/pull/10802#discussion_r684341804

pitrou commented on python/pyarrow/table.pxi:
```diff
@@ -956,6 +963,13 @@ cdef class RecordBatch(_PandasConvertible):
 
+    def drop_null(self):
+        """
+        Remove missing values from an RecordBatch.
```
I would say "a record batch" or "a RecordBatch".

pitrou replied:
Ping.

pitrou commented on python/pyarrow/table.pxi:
```diff
@@ -397,6 +397,13 @@ cdef class ChunkedArray(_PandasConvertible):
+    def drop_null(self):
+        """
+        Remove missing values from a chunked array.
+        See pyarrow.compute.drop_null for full usage.
```
"for full description" perhaps, because the function signature (hence usage) is trivial?

pitrou commented on python/pyarrow/tests/test_compute.py:
```diff
@@ -974,6 +974,82 @@ def test_take_null_type():
+def test_drop_null(ty, values):
+    arr = pa.array(values, type=ty)
+    result = arr.drop_null()
+    result.validate()
```
`result.validate(full=True)` is more thorough.

pitrou commented on cpp/src/arrow/compute/kernels/vector_selection_test.cc:
```diff
@@ -1734,5 +1735,551 @@ TEST(TestTake, RandomFixedSizeBinary) {
+
+TEST_F(TestDropNullKernel, DropNullBoolean) {
+  CheckDropNull(boolean(), "[true, false, true]", "[true, false, true]");
+  CheckDropNull(boolean(), "[null, false, true]", "[false, true]");
```
Ideally, should also check with an empty array and a nulls-only array.

pitrou commented on cpp/src/arrow/compute/kernels/vector_selection_test.cc:
```diff
@@ -1734,5 +1735,551 @@ TEST(TestTake, RandomFixedSizeBinary) {
+    std::generate(offsets.begin(), offsets.end(), [&] { return dist(gen); });
+    std::sort(offsets.begin(), offsets.end());
+    return offsets;
+  }
```
Can probably reuse `RandomArrayGenerator::Offsets` instead.

aocsa replied:
I tried, but the specific number of offsets doesn't match all the time with null_probability  input for RandomArrayGenerator::Offsets generator.

pitrou replied:
Hmm, I'm not sure I understand what you mean here. You should just pass a zero null_probability for offsets (the default).

aocsa replied:
I change this and you are right, I didn't see that the default null_probability was zero

pitrou commented on python/pyarrow/tests/test_compute.py:
```diff
@@ -974,6 +974,82 @@ def test_take_null_type():
+    expected = pa.table([pa.array(["a", "d"]),
+                         pa.array(["a", "d"]),
+                         pa.array(["a", "d"])],
+                        names=["a", "b", "c"])
```
Hmm, you could probably make the values different in each column, though this is already tested on the C++ side, so probably not very important.

pitrou reviewed (changes requested):
Thanks for doing this @aocsa ! It's especially nice that the tests are comprehensive.

pitrou commented:
One last comment: can you document the function in `docs/source/cpp/compute.rst`?

pitrou commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,219 @@ class TakeMetaFunction : public MetaFunction {
+    ("The output is populated with values from the input (Array, ChunkedArray, "
+     "RecordBatch, or Table) without the null values."
+     "Note that for the RecordBatch/Table cases, `drop_null` drops the full row if there "
+     "is any null."),
```
Ah, one other thing: as a convention, we wrap FunctionDoc descriptions with explicit newlines (`\n`). It would be good to conform to that here.

aocsa commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,219 @@ class TakeMetaFunction : public MetaFunction {
+// ----------------------------------------------------------------------
+// DropNull Implementation
+
+Status GetDropNullFilter(const Array& values, MemoryPool* memory_pool,
```
previous to this I was using `Result<...>` for this helper function, Niranda suggest me this way instead. So I think this is just a preference. Personally I prefer the use of `Result<...>`

pitrou replied:
Hmm, I'm curious why @nirandaperera suggested this. Returning `Result<...>` is the recommended convention in all newly written code (an exception is when the out-parameter is optional, and/or you have several of them).

nirandaperera commented:
Lol! Yes. I think that was a mistake from me! 😭

On Fri, Aug 13, 2021, 11:05 Antoine Pitrou ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In cpp/src/arrow/compute/kernels/vector_selection.cc
> <https://github.com/apache/arrow/pull/10802#discussion_r688584422>:
>
> > @@ -2146,6 +2147,219 @@ class TakeMetaFunction : public MetaFunction {
>    }
>  };
>
> +// ----------------------------------------------------------------------
> +// DropNull Implementation
> +
> +Status GetDropNullFilter(const Array& values, MemoryPool* memory_pool,
>
> Hmm, I'm curious why @nirandaperera <https://github.com/nirandaperera>
> suggested this. Returning Result<...> is the recommended convention in
> all newly written code (an exception is when the out-parameter is optional,
> and/or you have several of them).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/arrow/pull/10802#discussion_r688584422>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKS65PYEWOTPW62DJEI2KDT4UYCXANCNFSM5BAK5SZA>
> .
>

aocsa commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,219 @@ class TakeMetaFunction : public MetaFunction {
+
+Status CreateEmptyChunkedArray(std::shared_ptr<DataType> type, MemoryPool* memory_pool,
+                               std::shared_ptr<ChunkedArray>* output_array) {
+  std::vector<std::shared_ptr<Array>> new_chunks(1);  // Hard-coded 1 for now
```
Yes, it is allowed, but for consistency with Take:

https://github.com/apache/arrow/blob/5c5a0d63a42dc8d5ecab5996574c466f2e9c2ed5/cpp/src/arrow/compute/kernels/vector_selection.cc#L2001

And with  ChunkedArray::Make  

https://github.com/apache/arrow/blob/5c5a0d63a42dc8d5ecab5996574c466f2e9c2ed5/cpp/src/arrow/chunked_array.cc#L72

This is preferred to be consistent with the expected output results.

aocsa commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,219 @@ class TakeMetaFunction : public MetaFunction {
+    RETURN_NOT_OK(
+        CreateEmptyChunkedArray(values.type(), ctx->memory_pool(), &empty_array));
+    return empty_array;
+  }
```
yeap, https://github.com/apache/arrow/blob/2675cb94ca26126f21a93ed596347545f1e72928/cpp/src/arrow/compute/kernels/vector_selection.cc#L2181

pitrou replied:
Right, but I don't see the same line here in `DropNullChunkedArray` :-) Am I misunderstanding something?

aocsa replied:
Sure, my mistake

aocsa commented:
> One last comment: can you document the function in `docs/source/cpp/compute.rst`?

Thanks for the feedback @pitrou. 
I updated this PR :)

pitrou commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,219 @@ class TakeMetaFunction : public MetaFunction {
+    null_count += column->null_count();
+  }
+  if (null_count == 0) {
+    return RecordBatch::Make(batch.schema(), batch.num_rows(), batch.columns());
```
Ping here. If you pass this function a `const std::shared_ptr<RecordBatch>& batch` argument, you can just `return batch` here AFAICT.

pitrou replied:
Still worth addressing.

aocsa replied:
Thanks I didn't realize that this was not changed.

pitrou commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,211 @@ class TakeMetaFunction : public MetaFunction {
+  auto drop_null_filter =
+      std::make_shared<BooleanArray>(batch.num_rows(), dst, nullptr, 0, 0);
+  if (drop_null_filter->null_count() == batch.num_rows()) {
+    std::vector<std::shared_ptr<Array>> empty_batch(batch.num_columns());
```
No need to change this, but note we have `ArrayVector` and other similar aliases (such as `BufferVector` etc.) which may ease typing.

aocsa commented:
Thanks again @pitrou, just a reminder  this PR was updated. Main change related to DropNullTable

> ... is to get TableBatchReader to get a vector of record batches spanning the table, then call drop_null on each record batch and then Table::FromRecordBatches to get the resulting table.

https://github.com/apache/arrow/pull/10802#discussion_r687883363

pitrou commented on python/pyarrow/compute.py:
```diff
@@ -591,6 +591,38 @@ def take(data, indices, *, boundscheck=True, memory_pool=None):
     return call_function('take', [data, indices], options, memory_pool)
 
 
+def drop_null(data, memory_pool=None):
```
I'll remove this as the proper definition is auto-generated from the C++ metadata.

pitrou commented on cpp/src/arrow/compute/kernels/vector_selection.cc:
```diff
@@ -2146,6 +2147,167 @@ class TakeMetaFunction : public MetaFunction {
+  }
+  auto drop_null_filter =
+      std::make_shared<BooleanArray>(batch->num_rows(), dst, nullptr, 0, 0);
+  if (drop_null_filter->null_count() == batch->num_rows()) {
```
This condition is never satisfied, will fix.

pitrou commented on cpp/src/arrow/compute/kernels/vector_selection_test.cc:
```diff
@@ -1734,5 +1735,548 @@ TEST(TestTake, RandomFixedSizeBinary) {
+    auto schm = schema(fields);
+    std::vector<std::shared_ptr<ChunkedArray>> empty_table(fields.size());
+    auto col_a = std::make_shared<NullArray>(size);
+    auto col_b = std::make_shared<NullArray>(size);
```
These are the wrong column types. I will fix and add a validation.

pitrou commented:
Note the auto-generated Python  docstring isn't entirely accurate (see the `input` argument description):
```
Signature: pc.drop_null(input, *, memory_pool=None)
Docstring:
Drop nulls from the input.

The output is populated with values from the input (Array, ChunkedArray,
RecordBatch, or Table) without the null values.
For the RecordBatch and Table cases, `drop_null` drops the full row if
there is any null.

Parameters
----------
input : Array-like or scalar-like
    Argument to compute function
memory_pool : pyarrow.MemoryPool, optional
    If not passed, will allocate memory from the default memory pool.
File:      ~/arrow/dev/python/pyarrow/compute.py
Type:      function
```

I'll open a separate JIRA.

aocsa commented:
> Note the auto-generated Python docstring isn't entirely accurate (see the `input` argument description):
> 
> ```
> Signature: pc.drop_null(input, *, memory_pool=None)
> Docstring:
> Drop nulls from the input.
> 
> The output is populated with values from the input (Array, ChunkedArray,
> RecordBatch, or Table) without the null values.
> For the RecordBatch and Table cases, `drop_null` drops the full row if
> there is any null.
> 
> Parameters
> ----------
> input : Array-like or scalar-like
>     Argument to compute function
> memory_pool : pyarrow.MemoryPool, optional
>     If not passed, will allocate memory from the default memory pool.
> File:      ~/arrow/dev/python/pyarrow/compute.py
> Type:      function
> ```
> 
> I'll open a separate JIRA.

Ok, thanks @pitrou let me know in which jira ticket will be this  to follow up the updates.

pitrou commented:
It's in https://issues.apache.org/jira/browse/ARROW-13665
